### ISSUE TITLE
Memory leak after task raise exception

### ISSUE THREAD

**Issue by joaojunior** [celery/celery#6058]
# Checklist

- [ ] I have verified that the issue exists against the `master` branch of Celery.
- [ ] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
- [x] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [x] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [ ] I have verified that the issue exists against the `master` branch of Celery.
- [x] I have included the contents of ``pip freeze`` in the issue.
- [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [x] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [x] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [x] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [x] I have tried reproducing the issue on more than one operating system.
- [x] I have tried reproducing the issue on more than one workers pool.
- [x] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [x] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- #6009
- #5230
- #4809
- #6023 

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**:

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
# With broker rabbitmq
software -> celery:4.4.2 (cliffs) kombu:4.6.8 py:3.8.0
            billiard:3.6.3.0 py-amqp:2.5.2
platform -> system:Darwin arch:64bit
            kernel version:19.4.0 imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:pyamqp results:disabled

broker_url: 'amqp://guest:********@localhost:5672//'

# With broker redis
software -> celery:4.4.2 (cliffs) kombu:4.6.8 py:3.8.0
            billiard:3.6.3.0 redis:3.4.1
platform -> system:Darwin arch:64bit
            kernel version:19.4.0 imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:disabled

broker_url: 'redis://localhost:6379/0'

```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

* **Minimal Python Version**: N/A or Unknown
* **Minimal Celery Version**: N/A or Unknown
* **Minimal Kombu Version**: N/A or Unknown
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: N/A or Unknown
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
amqp==2.5.2
billiard==3.6.3.0
celery==4.4.2
kombu==4.6.8
pytz==2019.3
redis==3.4.1
vine==1.3.0
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
from celery import Celery

app = Celery('tasks', broker='pyamqp://guest:guest@localhost:5672') # for rabitmq broker
# app = Celery('tasks', broker='redis://localhost:6379/0') for refis broker

@app.task
def task_example():
    s = 'a' * 256 * (10**6) # Create a string of 256MB
    raise
```

</p>
</details>

# Expected Behavior

I will describe all the scenarios that I tested.
- Scenario 1: I ran the worker with the command `celery worker -A tasks -lINFO` and after this I executed the task with the command `python -c "from tasks import task_example; task_example.delay()"`.
- Scenario 2: I ran the worker with the command `celery worker -A tasks -lINFO -P threads` and after this I executed the task with the command `python -c "from tasks import task_example; task_example.delay()"`.
- Scenario 3: I ran the worker with the command `celery worker -A tasks -lINFO  --max-memory-per-child 125000` and after this I executed the task with the command `python -c "from tasks import task_example; task_example.delay()"`.
- Scenario 4: I ran the worker with the command `celery worker -A tasks -lINFO -P threads  --max-memory-per-child 125000` and after this I executed the task with the command `python -c "from tasks import task_example; task_example.delay()"`.

In all of this scenarios I expect that after run the task, celery will not continue consume the memory.

# Actual Behavior

Only in scenario 3, as the documentation explains, the celery not continue consume memory because celery replaced the child process.

On scenarios 1, 2, and 4 the celery keep the memory that the task allocate(256MB for each task ran in this case).

**Comment by asfaltboy** [comment]
Thanks @joaojunior , this is a very important finding!

Note: this be fixed in PR https://github.com/celery/celery/pull/6024 , would you try reproducing on latest celery master branch?

If it's not fixed, then this leak SO answer shows how some tools can be useful in tracing down the culprit: https://stackoverflow.com/a/61260839/484127
And, we'll need to check if this leak also goes back to 4.0 or an earlier 3.x version.

**Comment by joaojunior** [comment]
@asfaltboy I saw your comments only now. I will try with the code of the PR and also with the master and I return here with the results.

**Comment by auvipy** [comment]
> @asfaltboy I saw your comments only now. I will try with the code of the PR and also with the master and I return here with the results.

did you check it?

**Comment by auvipy** [comment]
> Note: this could be fixed in PR #6024 , would you try reproducing on latest celery master branch?

closing for now as per this. will revisit after further report
